### PR TITLE
Python: Fix some problems in TaintKind useage

### PR DESCRIPTION
**Pull request by RasmusWL**
See details in commit messages.

### CODE REVIEW

**Comment by RasmusWL** [comment]
Uhhhh, I actually forgot to highlight a quite important detail :grimacing: 

Any `import semmle.python.web.HttpRequest` would also import `UntrustedStringKind` (transitively). By removing this feature, we might actually make queries stop giving results, if the query does not explicitly contain `import semmle.python.security.strings.Untrusted`, since there would be no non-abstract subclass of `ExternalStringKind`.

I'll make an update in the change-notes, but don't know if we need more than that?

**Comment by yoff** [comment]
It seems to me that commit 1 is unrelated to commits 2 and 3? The former is standardising some web libraries while the latter are cleaning up our taint tracking labels. Small as they are, I would probably have expected them to inhabit separate PRs.

In commit 1, before and after look alpha-equivalent to me, but I will take your word for it that UntrustedStringKind is more likely to cause conflicts than ExternalStringKind.

**Comment by yoff** [comment]
Would queries just fail silently without some error?

**Comment by RasmusWL** [comment]
> It seems to me that commit 1 is unrelated to commits 2 and 3? The former is standardising some web libraries while the latter are cleaning up our taint tracking labels. Small as they are, I would probably have expected them to inhabit separate PRs.

Yes, I am sometimes lazy and group things together that are related but could be split into multiple PRs.

> In commit 1, before and after look alpha-equivalent to me, but I will take your word for it that UntrustedStringKind is more likely to cause conflicts than ExternalStringKind.

If using `instanceof UntrustedStringKind`, it wouldn't be possible to become a source of `ExternalXmlString` (or any other custom TaintKind). Does that clear it up? :blush: 

https://github.com/github/codeql/blob/453de6bf4e1f26be7134e3a6d885f2d0f93c2235/python/ql/src/semmle/python/security/injection/Xml.qll#L49-L51

> Would queries just fail silently without some error?

If by "fail" you mean "not produce any results", then yes

File: python/ql/src/semmle/python/security/strings/Basic.qll [hunk 1]
```diff
@@ -107,7 +107,11 @@ private predicate os_path_join(ControlFlowNode fromnode, CallNode tonode) {
     tonode.getAnArg() = fromnode
 }
 
-/** A kind of "taint", representing a dictionary mapping str->"taint" */
-class StringDictKind extends DictKind {
+/**
+ * A kind of "taint", representing a dictionary mapping str->"taint"
+ *
+ * DEPRECATED: Use `ExternalStringKind` instead.
```
Comment (tausbn) [inline python/ql/src/semmle/python/security/strings/Basic.qll:113]:
```suggestion
 * DEPRECATED: Use `ExternalStringDictKind` instead.
```
... Unless I'm mistaken.

Comment (RasmusWL) [reply python/ql/src/semmle/python/security/strings/Basic.qll]: :grimacing:

**Comment by yoff** [comment]
Yes, so it seems to me that the importance of switching from `UntrustedStringKind` to `ExternalStringKind ` can only be understood in terms of the surrounding programming. That is, the programming discipline for reusing taint tracking definitions is via inheritance and a set of useful libraries has standardised on inheriting from exactly `ExternalStringKind`. This PR is bringing `bottle.ql` in line with that programming discipline. Is that correctly understood?

**Comment by yoff** [comment]
Thanks for the out-of-band explanation, happy to merge now.

**Pull request merged**